Deploy more smart pointers in HTMLSelectElement.cpp.

AlexsanderDamaceno opened the pull request:
#### f6d5526e6123330465ed88dc52d6325f09e62c1a
<pre>
Deploy more smart pointers in HTMLSelectElement.cpp.
<a href="https://bugs.webkit.org/show_bug.cgi?id=277479.">https://bugs.webkit.org/show_bug.cgi?id=277479.</a>

Reviewed by Ryosuke Niwa.

Deploy more smart pointers in HTMLSelectElement.cpp.

* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::setValue):
(WebCore::HTMLSelectElement::optionToSelectFromChildChangeScope):
(WebCore::HTMLSelectElement::setLength):
(WebCore::HTMLSelectElement::recalcListItems const):
(WebCore::HTMLSelectElement::searchOptionsForValue const):

Canonical link: <a href="https://commits.webkit.org/281750@main">https://commits.webkit.org/281750@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be8cdb1923d8f8da7ec43471d2060edeaa6dfc22

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40168 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13385 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64740 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11356 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62939 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47844 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11631 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49171 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7883 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62843 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37406 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52679 "Passed tests") | [⏳ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/API-Tests-WPE-EWS "Waiting to run tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34093 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9906 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10269 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55931 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10203 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66469 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4753 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56539 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4774 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52647 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56720 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13584 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3939 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35973 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37055 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38148 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36800 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -385,19 +385,19 @@ CompletionHandlerCallingScope HTMLSelectElement::optionToSelectFromChildChangeSc
 
     auto getLastSelectedOption = [](HTMLOptGroupElement& optGroup) -> HTMLOptionElement* {
-        for (auto* option = Traversal<HTMLOptionElement>::lastChild(optGroup); option; option = Traversal<HTMLOptionElement>::previousSibling(*option)) {
+        for (RefPtr option = Traversal<HTMLOptionElement>::lastChild(optGroup); option; option = Traversal<HTMLOptionElement>::previousSibling(*option)) {
```
This should not be needed. `selectedWithoutUpdate()` is a trivial getter.

cdumez commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -385,19 +385,19 @@ CompletionHandlerCallingScope HTMLSelectElement::optionToSelectFromChildChangeSc
     RefPtr<HTMLOptionElement> optionToSelect;
     if (change.type == ChildChange::Type::ElementInserted) {
-        if (auto* option = dynamicDowncast<HTMLOptionElement>(*change.siblingChanged)) {
+        if (RefPtr option = dynamicDowncast<HTMLOptionElement>(*change.siblingChanged)) {
```
ditto.

cdumez commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -519,7 +519,7 @@ ExceptionOr<void> HTMLSelectElement::setLength(unsigned newLength)
         size_t optionIndex = 0;
         for (auto& item : items) {
-            auto* option = dynamicDowncast<HTMLOptionElement>(*item);
+            RefPtr option = dynamicDowncast<HTMLOptionElement>(*item);
```
Should not be needed.

AlexsanderDamaceno committed:
Deploy more smart pointers in HTMLSelectElement.cpp.
https://bugs.webkit.org/show_bug.cgi?id=277479.

Reviewed by Ryosuke Niwa.

Deploy more smart pointers in HTMLSelectElement.cpp.

* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::setValue):
(WebCore::HTMLSelectElement::optionToSelectFromChildChangeScope):
(WebCore::HTMLSelectElement::setLength):
(WebCore::HTMLSelectElement::recalcListItems const):
(WebCore::HTMLSelectElement::searchOptionsForValue const):

Canonical link: https://commits.webkit.org/281750@main
